Project binding

ftomassetti opened the pull request:
Implement project binding and a set of supporting changes.

Addition:
- integration tests
- manual tests
- more flexible model-server (in-memory storage, possibility to load/save dumps)
- storing binding configuration

It will need to be further polished and I am sure there will be bug to fix, but I think it could be better to reconciliate before further diverging

slisson commented:
When I add a repository in the cloud view and restart MPS, the list of repositories is empty. The headless MPS running in the kubernetes cluster doesn't bind the modules store on the model server, probably also because the list of repositories is empty and not initialized with the default one.

ftomassetti commented:
The list of repositories should be loaded from the file `.mps/cloudResources.xml`. If that file does not exist the list is empty (there is no default repository added).

If you add the repository, close MPS and re-open it, and the repository is not there, this is a bug and should be fixed. The file cloudResources.xml is saved when MPS decides too (i.e., we do not control explicitly when it happens: we change the state but I think it is flushed to file only on exit).

Regarding the fact that before we add a default entry present, I thought it would make sense to remove it (provided the cloudResources.xml mechanism work), but if you prefer we can add that entry when the file cloudResources.xml is not present or it is present but no repositories are specified (I am not sure we can easily distinguish the two cases).

slisson commented:
Adding the repository passed as a command line parameter is enough. See https://github.com/modelix/modelix/blob/mps2020.3/run-ui-server.sh#L3

I opened MPS, added a repository and restarted MPS. Then the list was empty and there also was no `.mps/cloudResources.xml` file.

ftomassetti commented:
> I opened MPS, added a repository and restarted MPS. Then the list was empty and there also was no `.mps/cloudResources.xml` file.

I tried this on my side: I have open a project, I added a cloud repository, closed the project and the file cloudResources.xml was created with this content:

```
<?xml version="1.0" encoding="UTF-8"?>
<project version="4">
  <component name="CloudResources">
    <option name="cloudRepositories">
      <list>
        <option value="http://localhost:8899/" />
      </list>
    </option>
  </component>
</project>
```

The code that adds this configuration to the state is not triggered automatically when adding an element into CloudRepositories, but it is invoked by the `AddCloudRepository` action (at this line: `PersistedBindingConfiguration.getInstance(this.project).addRepository(cloudRepository);`)

Did you add the project using that action or in some other way I did not consider?

The only reason for this not to work that I can think of is that for some reason the component state is not saved. Maybe because of the way MPS was closed? Or because there were more projects open at the same time? Any log that could help?
If no other solution can be found we could drop the usage of the component and save the file directly.

The component handling the persistence of the configuration is `CloudResourcesConfigurationComponent` which extends `PersistentStateComponent`.

I will take another reading at https://jetbrains.org/intellij/sdk/docs/basics/persisting_state_of_components.html 

> Adding the repository passed as a command line parameter is enough. See https://github.com/modelix/modelix/blob/project-binding-mps2020.3/run-ui-server.sh#L3

Regarding this point, are you suggesting we make possible to add a default cloud repository when starting MPS from the command line, like you do for the ui-server?

slisson commented:
> Or because there were more projects open at the same time?

Yes, that was the reason. I had multiple projects loaded before. Now it works when I only open the modelix project.

> Regarding this point, are you suggesting we make possible to add a default cloud repository when starting MPS from the command line, like you do for the ui-server?

Yes, I need a command line parameter that MPS then automatically connects to.

ftomassetti commented:
> > Or because there were more projects open at the same time?
> 
> Yes, that was the reason. I had multiple projects loaded before. Now it works when I only open the modelix project.

This seems a problem to be investigated further. I have created this issue to handle it https://github.com/modelix/modelix/issues/59 

> Yes, I need a command line parameter that MPS then automatically connects to.

Ok, I will work on a patch to ProjectPlugin, so that when the plugin is loaded I will look into System property and, if they are present, I will add a cloud repository to the project.

Should this project also be stored in the cloudResources.xml as for the ones manually added?
Do we want an option also to do the automated binding of the physical project to a cloud project?

ftomassetti commented:
I have added this:

<img width="1097" alt="Screenshot 2021-01-25 at 16 49 14" src="https://user-images.githubusercontent.com/439078/105729253-50e0aa00-5f2d-11eb-898d-c5a52af8ea44.png">

However, I am not sure how to test it. To run Modelix headless do we have a task? And then I could I check the cloud repo is present? I want to be sure I would test as you would do it, so that I get it working for you

The way it should work is by adding all cloud repos specified in the system property "modelix.cloud.repos", using comma to separate multiple URLs

slisson commented:
I fixed it.
What I do to test it is
- Install docker desktop and enable kubernetes
- `git clean -xdf && ./gradlew && ./docker-build-all.sh && ./kubernetes-apply-local.sh`
- `watch kubectl get pods`
- Wait for pod/ui-... to become ready. If it doesn't turn from `0/1` to `1/1` after a few minutes, it doesn't work.

ftomassetti commented:
Thank you for handling it!
Do you think is there anything else I should get fixed before merging?

slisson commented:
No, can be merged now.
